File: C#/KnowItTest/Passenger.cs
public class Passenger
{
    public double TotalBill { get; private set; }
    private Transportation? currentTransportation;
    private List<DateTime> tripDates = new List<DateTime>();

    public void CheckIn(Transportation transportation, DateTime date)
    {
        currentTransportation = transportation;
        tripDates.Add(date);
    }

    public void CheckOut(DateTime date, PublicHoliday publicHoliday)
    {
        if (currentTransportation != null)
        {
            DateTime tripStartDate = tripDates.LastOrDefault();
            DateTime tripEndDate = date;

            int tripDuration = (int)(tripEndDate - tripStartDate).TotalDays;
            int highestPrice = 0;

            if (publicHoliday.Holidays.Contains(tripEndDate.Date))
            {
                highestPrice = 0;
            }
            else
            {
                for (int i = 0; i <= tripDuration; i++)
                {
                    DateTime currentDay = tripStartDate.AddDays(i);
                    int dayPrice = currentTransportation.Prices[currentDay.DayOfWeek];
                    if (dayPrice > highestPrice)
                    {
                        highestPrice = dayPrice;
                    }
                }
            }

            TotalBill += highestPrice;

            if (TotalBill > 200)
            {
                TotalBill = 200;
            }

            // Reset the current transportation
            currentTransportation = null;
        }
        else
        {
            Console.WriteLine("Checkout must be paired with a previous Checkin.");
        }
    }
}

File: Java/Passenger.java
import java.time.DayOfWeek;
import java.time.LocalDate;
import java.util.*;

public class Passenger {
    private double totalBill;
    private Transportation currentTransportation;
    private List<LocalDate> tripDates = new ArrayList<>();

    public void checkIn(Transportation transportation, LocalDate date) {
        currentTransportation = transportation;
        tripDates.add(date);
    }

    public void checkOut(LocalDate date, PublicHoliday publicHoliday) {
        if (currentTransportation != null) {
            LocalDate tripStartDate = tripDates.get(tripDates.size() - 1);
            LocalDate tripEndDate = date;

            int tripDuration = (int) tripStartDate.until(tripEndDate).getDays();
            int highestPrice = 0;

            if (publicHoliday.getHolidays().contains(tripEndDate)) {
                highestPrice = 0;
            } else {
                for (int i = 0; i <= tripDuration; i++) {
                    LocalDate currentDay = tripStartDate.plusDays(i);
                    int dayPrice = currentTransportation.getPrices().get(currentDay.getDayOfWeek());
                    if (dayPrice > highestPrice) {
                        highestPrice = dayPrice;
                    }
                }
            }

            totalBill += highestPrice;

            if (totalBill > 200) {
                totalBill = 200;
            }

            // Reset the current transportation
            currentTransportation = null;
        } else {
            System.out.println("Checkout must be paired with a previous Checkin.");
        }
    }

    public double getTotalBill() {
        return totalBill;
    }
}